### PR TITLE
feat(expression router) add validation to creating predicates

### LINKED ISSUE #3822: Expression Router: Validate Predicates when creating predicates
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem Statement

Now we could create invalid expressions in `NewPredicate` in expression generators. We should check the validity of predicate to avoid invalid expressions appearing in kong routes, that will cause updating Kong configurations stuck.

### Proposed Solution

- Check validity of predicates and update signature of `NewPredicate` function to check:
  - if LHS and RHS has the matching type (String - String, Int - Int, Others are invalid)
  - If operator can be applied to the type of sides (e.g: ,`^=`(prefix match) only for strings)

### Additional information

Sub issue of #3730 

### Acceptance Criteria

- [ ] `NewPredicate` function and `NewPredicate***` functions to create predicate for specific fields returns error when invalid parameters are put in

### PR DESCRIPTION
**Pull request by randmonkey**
**What this PR does / why we need it**:

Adds validation on creating ATC predicates in `NewPredicate`.

**Which issue this PR fixes**:

fixes #3822.

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

~- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR~ No user facing part changed.

### CODE REVIEW

File: internal/dataplane/parser/atc/predicate.go [hunk 1]
```diff
@@ -115,15 +116,36 @@ func (p Predicate) IsEmpty() bool {
...
+	ErrorTypeNotMatch    = errors.New("type does not match on sides of predicate")
+	ErrorOperatorInvalid = errors.New("operator is not valid for the types of sides of predicate")
+)
+
 // NewPredicate generates a single predicate.
-// TODO: check validity of LHS, op and RHS:
-// https://github.com/Kong/kubernetes-ingress-controller/issues/3822.
-func NewPredicate(lhs LHS, op BinaryOperator, rhs Literal) Predicate {
+func NewPredicate(lhs LHS, op BinaryOperator, rhs Literal) (Predicate, error) {
+	// check for predicates on string fields
```
Comment (programmer04) [inline internal/dataplane/parser/atc/predicate.go:126]:
```suggestion
	// Check for predicates on string fields.
```

File: internal/dataplane/parser/atc/predicate.go [hunk 2]
```diff
@@ -115,15 +116,36 @@ func (p Predicate) IsEmpty() bool {
...
+	if lhs.FieldType() == FieldTypeString {
+		if rhs.Type() != LiteralTypeString {
+			return Predicate{}, ErrorTypeNotMatch
+		}
+		if op == OpGreaterThan || op == OpGreaterEqual || op == OpLessThan || op == OpLessEqual {
+			return Predicate{}, ErrorOperatorInvalid
+		}
+	}
+	// check for predicates on integer fields
```
Comment (programmer04) [inline internal/dataplane/parser/atc/predicate.go:135]:
```suggestion
	// Check for predicates on integer fields.
```

**Review by programmer04 — COMMENTED** [review]
Could you add test cases that explicitly cover this functionality?

File: internal/dataplane/parser/atc/predicate.go [hunk 3]
```diff
@@ -115,15 +116,36 @@ func (p Predicate) IsEmpty() bool {
 	return p.value == nil
 }
 
+var (
+	ErrorTypeNotMatch    = errors.New("type does not match on sides of predicate")
+	ErrorOperatorInvalid = errors.New("operator is not valid for the types of sides of predicate")
```
Comment (pmalek) [inline internal/dataplane/parser/atc/predicate.go:121]: These should be called `Err...` per the [convention](https://github.com/golang/go/wiki/Errors#naming).

I've introduced the `errname` linter to enforce this https://github.com/Kong/kubernetes-ingress-controller/pull/4811

File: internal/dataplane/parser/atc/predicate.go (hunk 1 above)
Comment (randmonkey) [reply internal/dataplane/parser/atc/predicate.go]: Done

File: internal/dataplane/parser/atc/predicate.go (hunk 2 above)
Comment (randmonkey) [reply internal/dataplane/parser/atc/predicate.go]: done

File: internal/dataplane/parser/atc/predicate.go (hunk 3 above)
Comment (randmonkey) [reply internal/dataplane/parser/atc/predicate.go]: updated

**Comment by randmonkey** [comment]
> Could you add test cases that explicitly cover this functionality?

Added unit tests for `NewPredicate`.

**Pull request merged**